File: flint/execution/borrow_ledger.py
"""BorrowLedger — Jupiter Perps borrow history + paid-borrow ledger.

D-2.1.b Step 6 (Wave 2). Owns the per-market `BorrowSnapshot` history
plus the running `total_paid` counter and the per-trade payment ledger
that BacktestContext keeps for tearsheet attribution.

The two halves are colocated because both are about the same Jupiter
Perps product: rates come in via `record(snapshot)`, and when a
position is realized the borrow cost is computed from
`cumulative_at(market, ts)` and pushed into the payment ledger via
`record_payment(...)` for downstream attribution.
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional, Tuple

from ..models import BorrowSnapshot


class BorrowLedger:
    __slots__ = ("_history", "_payments", "total_paid")

    def __init__(self) -> None:
        self._history: Dict[str, List[BorrowSnapshot]] = {}
        self._payments: List[Dict[str, Any]] = []
        self.total_paid: float = 0.0

    # --- record ------------------------------------------------------------

    def record(self, snapshot: BorrowSnapshot) -> None:
        self._history.setdefault(snapshot.market, []).append(snapshot)

    def record_payment(self, payment: Dict[str, Any]) -> None:
        self._payments.append(payment)

    def add_paid(self, amount: float) -> None:
        self.total_paid += amount

    # --- read --------------------------------------------------------------

    def latest(self, market: str) -> Optional[float]:
        history = self._history.get(market)
        return history[-1].rate_hourly if history else None

    def recent(self, market: str, lookback: int = 24) -> List[Tuple[int, float]]:
        history = self._history.get(market, [])
        sliced = history[-lookback:] if lookback < len(history) else history
        return [(bs.ts, bs.rate_hourly) for bs in sliced]

    def cumulative_at(self, market: str, ts: int) -> Optional[float]:
        """Cumulative-rate value at or before `ts`. Snapshots in
        `_history` are append-ordered by arrival, which matches
        time order in the engine — same scan strategy as the
        pre-extraction code."""
        history = self._history.get(market, [])
        result: Optional[float] = None
        for bs in history:
            if bs.ts <= ts:
                result = bs.cumulative_rate
            else:
                break
        return result

    # --- raw access for legacy aliases -----------------------------------

    @property
    def history(self) -> Dict[str, List[BorrowSnapshot]]:
        return self._history

    @property
    def payments(self) -> List[Dict[str, Any]]:
        return self._payments

File: rust/src/engine/borrow_ledger.rs
//! BorrowLedger — Rust port of `flint/execution/borrow_ledger.py`.
//!
//! D-6.4-replay (Wave 5 Rust ledgers). Owns Jupiter Perps borrow-rate
//! history + the running `total_paid` counter + per-trade payment
//! ledger. Linear scan for `cumulative_at(ts)` mirrors the Python
//! impl — snapshots are append-ordered by arrival, which matches
//! time order in the engine.

use std::collections::HashMap;

#[derive(Debug, Clone, Copy)]
pub struct BorrowPoint {
    pub ts: i64,
    pub rate_hourly: f64,
    pub cumulative_rate: f64,
}

#[derive(Debug, Clone)]
pub struct BorrowPayment {
    pub market: String,
    pub ts: i64,
    pub cost: f64,
    pub cum_entry: f64,
    pub cum_exit: f64,
}

#[derive(Debug, Default, Clone)]
pub struct BorrowLedger {
    history: HashMap<String, Vec<BorrowPoint>>,
    payments: Vec<BorrowPayment>,
    pub total_paid: f64,
}

impl BorrowLedger {
    pub fn new() -> Self { Self::default() }

    pub fn record(&mut self, market: &str, ts: i64, rate_hourly: f64, cumulative_rate: f64) {
        self.history.entry(market.to_string()).or_default().push(BorrowPoint {
            ts, rate_hourly, cumulative_rate,
        });
    }

    pub fn record_payment(&mut self, p: BorrowPayment) {
        self.payments.push(p);
    }

    pub fn add_paid(&mut self, amount: f64) {
        self.total_paid += amount;
    }

    pub fn latest(&self, market: &str) -> Option<f64> {
        self.history.get(market).and_then(|h| h.last()).map(|p| p.rate_hourly)
    }

    pub fn recent(&self, market: &str, lookback: usize) -> Vec<(i64, f64)> {
        let h = match self.history.get(market) { Some(h) => h, None => return Vec::new() };
        let start = h.len().saturating_sub(lookback);
        h[start..].iter().map(|p| (p.ts, p.rate_hourly)).collect()
    }

    /// Cumulative-rate value at-or-before `ts`. Linear walk because
    /// snapshots arrive append-ordered and an early-exit on the
    /// first point past `ts` is the right behavior.
    pub fn cumulative_at(&self, market: &str, ts: i64) -> Option<f64> {
        let h = self.history.get(market)?;
        let mut out: Option<f64> = None;
        for pt in h {
            if pt.ts <= ts {
                out = Some(pt.cumulative_rate);
            } else {
                break;
            }
        }
        out
    }

    pub fn payments(&self) -> &[BorrowPayment] { &self.payments }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn latest_and_cumulative() {
        let mut bl = BorrowLedger::new();
        bl.record("JUP-PERP", 10, 0.0001, 0.005);
        bl.record("JUP-PERP", 20, 0.0002, 0.010);
        bl.record("JUP-PERP", 30, 0.00015, 0.015);
        assert_eq!(bl.latest("JUP-PERP"), Some(0.00015));
        assert!(bl.cumulative_at("JUP-PERP", 5).is_none());      // before all
        assert_eq!(bl.cumulative_at("JUP-PERP", 10), Some(0.005));
        assert_eq!(bl.cumulative_at("JUP-PERP", 25), Some(0.010));
        assert_eq!(bl.cumulative_at("JUP-PERP", 100), Some(0.015));
    }

    #[test]
    fn payments_and_total() {
        let mut bl = BorrowLedger::new();
        bl.record_payment(BorrowPayment {
            market: "JUP-PERP".to_string(), ts: 1, cost: 1.5,
            cum_entry: 0.001, cum_exit: 0.002,
        });
        bl.add_paid(1.5);
        assert_eq!(bl.total_paid, 1.5);
        assert_eq!(bl.payments().len(), 1);
        assert_eq!(bl.payments()[0].cost, 1.5);
    }

    #[test]
    fn recent_lookback_truncates() {
        let mut bl = BorrowLedger::new();
        for i in 0..50 {
            bl.record("J", i as i64, 0.0001 * i as f64, 0.0);
        }
        let r = bl.recent("J", 5);
        assert_eq!(r.len(), 5);
        assert_eq!(r[0].0, 45);
    }
}
